onyo/lib/utils.py

```diff
@@ -2,52 +2,75 @@
 
 import copy
 from pathlib import Path
-from typing import Dict
+from typing import TYPE_CHECKING
 
 from ruamel.yaml import YAML, scanner  # pyre-ignore[21]
 
 from onyo.lib.consts import PSEUDO_KEYS, RESERVED_KEYS
 from onyo.lib.ui import ui
 
+if TYPE_CHECKING:
+    from typing import Dict, Set
+
 
 def deduplicate(sequence: list | None) -> list | None:
-    r"""Get a deduplicated list, while preserving order.
+    r"""Deduplicate a list and preserve its order.
+
+    The first occurrence of a value is kept. All later occurrences are
+    discarded.
 
-    For ease of use, accepts `None` (and returns it in that case).
+    For convenience, also accepts ``None`` and returns ``None`` in that case.
```

Comment:
Good catch.

This stackoverflow comment has a nice overview that helped me think about this: https://stackoverflow.com/a/72805661

Basically, I like the idea of staying "imperative" whenever we can. However, to rewrite this in the imperative I think would be confusing (and imply that we're telling the user to do this, rather than describing what the function does)

> For convenience, also accepts ``None`` and returns ``None`` in that case.

to

> For convenience, also accept ``None`` and return ``None`` in that case.

That just feels weird.

The [pydocstyle project](https://www.pydocstyle.org/en/stable/error_codes.html) seems to only check for this on the first line. Which makes a lot of sense to me as a rule.

The short summary is imperative. But the extended summary is meant to explain a variety of things. Depending on the topic, the style will need to adapt.